db: data race with manifest.L0Organizer

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/19304138?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/19304138?buildTab=artifacts#meta) on refs/heads/master @ [c0b18bc8970e](https://github.com/cockroachdb/pebble/commits/c0b18bc8970e):


```
        db1.Set("ltfiguyav", "givpi") // <nil> #1037
        db1.SingleDelete("hoedxsh@2", false /* maybeReplaceDelete */) // <nil> #1038
        db1.RangeKeyUnset("fygfkcwsj", "ialfiyohhcd", "@7") // <nil> #1039
        db1.SingleDelete("bxbgtwq@5", false /* maybeReplaceDelete */) // <nil> #1040
        db1.DeleteRange("ltfiguyav", "oqyvflhu") // <nil> #1041
        db1.Merge("aswdunglrchp@5", "bhhyudiiszlkjiti") // <nil> #1042
        db1.Set("fbqryqypc", "wpql") // <nil> #1043
        db1.Merge("cplcdi@6", "keng") // <nil> #1044
        db1.Delete("bxbgtwq@4") // <nil> #1045
        db1.Set("dqccfcxuax", "ivepevotxj") // <nil> #1046
        db1.Merge("tpnhsbn", "assjfrtwobwbdt") // <nil> #1047
        db1.LogData("n") // <nil> #1048
        // 04:08:50.022 INFO: possible API misuse: ineffectual SINGLEDEL (key="xihzpiiott@3")
        // 04:08:50.024 INFO: [JOB 38] compacted(elision-only) L6 [000021] (3.2KB) Score=0.00 + L6 [] (0B) Score=0.00 -> L6 [000025] (1.4KB), in 0.0s (0.0s total), output rate 208KB/s
        snap9.Get("wzfowhh@1") // [""] pebble: not found #1049
        snap9.Get("kvshykocdfgs") // [""] pebble: not found #1050
        db1.Set("uvdmbswxx@9", "dzpcgesfewkp") // <nil> #1051
        db1.Delete("hoedxsh@2") // <nil> #1052
        db1.Delete("owquuxvdwsq@1") // <nil> #1053
        db1.Delete("dwoadyz@1") // <nil> #1054
        db1.Delete("bqex@5") // <nil> #1055
        db1.RangeKeySet("ntpbp", "ujvmmflmf", "@9", "oapwbvv") // <nil> #1056
        db1.RangeKeyUnset("kgudeihxl", "xsgyuijvc", "@6") // <nil> #1057
        db1.SingleDelete("vjlyodk@7", false /* maybeReplaceDelete */) // <nil> #1058
        db1.RangeKeySet("ewgw", "pfusfqpzvq", "@5", "qrkhvfbsyczerqbufur") // <nil> #1059
        db1.Delete("ffhkc") // <nil> #1060
        db1.Delete("ltfiguyav@6") // <nil> #1061
        db1.RangeKeyUnset("ntpbp", "owquuxvdwsq", "") // <nil> #1062
        db1.Get("oqyvflhu@6") // [""] pebble: not found #1063
        db1.Delete("yepfpcoemt@5") // <nil> #1064
        db1.LogData("cb") // <nil> #1065
        batch7 = db1.NewBatch() #1066
        batch7.Merge("bqex@5", "ysvnujdfjl") // <nil> #1068
        batch7.Merge("vnoz@9", "zhfky") // <nil> #1075
        // 04:08:50.027 INFO: [JOB 35] flushed 1 memtable (2.3KB) to L0 [000026] (1.7KB), in 0.0s (0.0s total), output rate 90KB/s
        // 04:08:50.027 INFO: [JOB 39] flushing 1 ingested table
        snap9.Get("xihzpiiott@2") // [""] pebble: not found #1067
        snap9.Get("tpnhsbn") // [""] pebble: not found #1069
        db1.Delete("wzfowhh@1") // <nil> #1070
        db1.SingleDelete("dwoadyz@1", false /* maybeReplaceDelete */) // <nil> #1071
        db1.Set("rsyypv@9", "riuzbyqrbjsgub") // <nil> #1072
        db1.RangeKeySet("jmqyjo", "omjr", "@8", "") // <nil> #1073
        db1.Set("cplcdi@6", "zdkcdksthsmhc") // <nil> #1074
        // 04:08:50.035 INFO: [JOB 38] sstable deleted 000021
        // 04:08:50.035 INFO: [JOB 35] WAL deleted 000018
        // 04:08:50.037 INFO: [JOB 41] ingesting: sstable created 000027
        // 04:08:50.038 INFO: [JOB 42] WAL created 000028
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/250403-040838.6934054690118/standard-014 --try-to-reduce -v
--- FAIL: TestMeta/execution/standard-014 (0.73s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1743653318693259594 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: PEBBLE-379

--- a/internal/keyspan/keyspanimpl/level_iter_test.go
+++ b/internal/keyspan/keyspanimpl/level_iter_test.go
@@ -314,8 +314,9 @@ func TestLevelIterEquivalence(t *testing.T) {
 			b.AddedTables[6] = amap
 			l0Organizer := manifest.NewL0Organizer(base.DefaultComparer, 0 /* flushSplitBytes */)
 			emptyVersion := manifest.NewInitialVersion(base.DefaultComparer)
-			v, err := b.Apply(emptyVersion, l0Organizer, 0)
+			v, err := b.Apply(emptyVersion, 0)
 			require.NoError(t, err)
+			l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(b, v), v)
 			levelIter.Init(
 				context.Background(),
 				keyspan.SpanIterOptions{}, base.DefaultComparer.Compare, tableNewIters,

--- a/internal/manifest/l0_sublevels.go
+++ b/internal/manifest/l0_sublevels.go
@@ -18,10 +18,6 @@ import (
 	"github.com/cockroachdb/pebble/internal/invariants"
 )
 
-// errInvalidL0SublevelsOpt is for use in addL0Files when the incremental
-// sublevel generation optimization failed, and newL0Sublevels must be called.
-var errInvalidL0SublevelsOpt = errors.New("pebble: L0 sublevel generation optimization cannot be used")
-
 // Intervals are of the form [start, end) with no gap between intervals. Each
 // file overlaps perfectly with a sequence of intervals. This perfect overlap
 // occurs because the union of file boundary keys is used to pick intervals.
@@ -314,9 +310,7 @@ func newL0Sublevels(
 	// Initialize minIntervalIndex and maxIntervalIndex for each file, and use that
 	// to update intervals.
 	for f := iter.First(); f != nil; f = iter.Next() {
-		if err := s.addFileToSublevels(f, false /* checkInvariant */); err != nil {
-			return nil, err
-		}
+		s.addFileToSublevels(f)
 	}
 	// Sort each sublevel in increasing key order.
 	for i := range s.levelFiles {
@@ -409,51 +403,20 @@ func mergeIntervals(
 	return result, oldToNewMap
 }
 
-// addL0Files incrementally builds a new l0Sublevels for when the only change
-// since the receiver l0Sublevels was an addition of the specified tables, with
-// no L0 deletions. The common case of this is an ingestion or a flush. These
-// files can "sit on top" of existing sublevels, creating at most one new
-// sublevel for a flush (and possibly multiple for an ingestion), and at most
-// 2*len(files) additions to s.orderedIntervals. No files must have been deleted
-// from L0, and the added files must all be newer in sequence numbers than
-// existing files in l0Sublevels. The levelMetadata parameter corresponds to the
-// new L0 post addition of files. This method is meant to be significantly more
-// performant than newL0Sublevels.
-//
-// Note that this function can only be called once on a given receiver; it
-// appends to some slices in s which is only safe when done once. This is okay,
-// as the common case (generating a new l0Sublevels after a flush/ingestion) is
-// only going to necessitate one call of this method on a given receiver. The
-// returned value, if non-nil, can then have [*l0Sublevels.addL0Files] called on
-// it again, and so on. If [errInvalidL0SublevelsOpt] is returned as an error,
-// it means the optimization could not be applied (i.e. files added were older
-// than files already in the sublevels, which is possible around ingestions and
-// in tests). Eg. it can happen when an ingested file was ingested without
-// queueing a flush since it did not actually overlap with any keys in the
-// memtable. Later on the memtable was flushed, and the memtable had keys
-// spanning around the ingested file, producing a flushed file that overlapped
-// with the ingested file in file bounds but not in keys. It's possible for that
-// flushed file to have a lower LargestSeqNum than the ingested file if all the
-// additions after the ingestion were to another flushed file that was split
-// into a separate sstable during flush. Any other non-nil error means
-// [l0Sublevels] generation failed in the same way as [newL0Sublevels] would
-// likely fail.
-func (s *l0Sublevels) addL0Files(
-	addedTables map[base.FileNum]*TableMetadata,
-	flushSplitMaxBytes int64,
-	levelMetadata *LevelMetadata,
-) (*l0Sublevels, error) {
+func (s *l0Sublevels) canUseAddL0Files(
+	addedTables map[base.FileNum]*TableMetadata, levelMetadata *LevelMetadata,
+) (filesToAddInOrder []*TableMetadata, ok bool) {
 	if s.addL0FilesCalled {
 		if invariants.Enabled {
 			panic("addL0Files called twice on the same receiver")
 		}
-		return nil, errInvalidL0SublevelsOpt
+		return nil, false
 	}
 	if s.levelMetadata.Len()+len(addedTables) != levelMetadata.Len() {
 		if invariants.Enabled {
 			panic("levelMetadata mismatch")
 		}
-		return nil, errInvalidL0SublevelsOpt
+		return nil, false
 	}
 
 	// addL0Files only works when the files we are adding match exactly the last
@@ -465,12 +428,40 @@ func (s *l0Sublevels) addL0Files(
 		if addedTables[t.FileNum] == nil {
 			// t is an existing table that sorts after some of the new tables
 			// (specifically the ones we haven't yet seen).
-			return nil, errInvalidL0SublevelsOpt
+			return nil, false
 		}
 		files[i] = t
 		t = iter.Prev()
 	}
-	// Note: files now contains addedTables in increasing seqnum order.
+	return files, true
+}
+
+// addL0Files incrementally builds a new l0Sublevels for when the only change
+// since the receiver l0Sublevels was an addition of the specified tables, with
+// no L0 deletions. The common case of this is an ingestion or a flush. These
+// files can "sit on top" of existing sublevels, creating at most one new
+// sublevel for a flush (and possibly multiple for an ingestion), and at most
+// 2*len(files) additions to s.orderedIntervals. No files must have been deleted
+// from L0, and the added files must all be newer in sequence numbers than
+// existing files in l0Sublevels. The levelMetadata parameter corresponds to the
+// new L0 post addition of files. This method is meant to be significantly more
+// performant than newL0Sublevels.
+//
+// This function is intended to be called with the result of canUseAddL0Files(),
+// which is the list of new L0 tables in increasing L0 order.
+//
+// Note that this function can only be called once on a given receiver; it
+// appends to some slices in s which is only safe when done once. This is okay,
+// as the common case (generating a new l0Sublevels after a flush/ingestion) is
+// only going to necessitate one call of this method on a given receiver. The
+// returned value, if non-nil, can then have [*l0Sublevels.addL0Files] called on
+// it again, and so on.
+func (s *l0Sublevels) addL0Files(
+	files []*TableMetadata, flushSplitMaxBytes int64, levelMetadata *LevelMetadata,
+) *l0Sublevels {
+	if s.addL0FilesCalled {
+		panic("addL0Files called twice on the same receiver")
+	}
 	s.addL0FilesCalled = true
 
 	// Start with a shallow copy of s.
@@ -618,9 +609,7 @@ func (s *l0Sublevels) addL0Files(
 	// Update interval indices for new files.
 	for i, f := range files {
 		f.L0Index = s.levelMetadata.Len() + i
-		if err := newVal.addFileToSublevels(f, true /* checkInvariant */); err != nil {
-			return nil, err
-		}
+		newVal.addFileToSublevels(f)
 		updatedSublevels = append(updatedSublevels, f.SubLevel)
 	}
 
@@ -658,16 +647,14 @@ func (s *l0Sublevels) addL0Files(
 	newVal.flushSplitUserKeys = nil
 	newVal.calculateFlushSplitKeys(flushSplitMaxBytes)
 	newVal.Check()
-	return newVal, nil
+	return newVal
 }
 
 // addFileToSublevels is called during l0Sublevels generation, and adds f to the
 // correct sublevel's levelFiles, the relevant intervals' files slices, and sets
 // interval indices on f. This method, if called successively on multiple files,
-// _must_ be called on successively newer files (by seqnum). If checkInvariant
-// is true, it could check for this in some cases and return
-// [errInvalidL0SublevelsOpt] if that invariant isn't held.
-func (s *l0Sublevels) addFileToSublevels(f *TableMetadata, checkInvariant bool) error {
+// _must_ be called on successively newer files (by seqnum).
+func (s *l0Sublevels) addFileToSublevels(f *TableMetadata) {
 	// This is a simple and not very accurate estimate of the number of
 	// bytes this SSTable contributes to the intervals it is a part of.
 	//
@@ -680,10 +667,8 @@ func (s *l0Sublevels) addFileToSublevels(f *TableMetadata, checkInvariant bool) 
 	for i := f.minIntervalIndex; i <= f.maxIntervalIndex; i++ {
 		interval := &s.orderedIntervals[i]
 		if len(interval.files) > 0 {
-			if checkInvariant && interval.files[len(interval.files)-1].LargestSeqNum > f.LargestSeqNum {
-				// We are sliding this file "underneath" an existing file. Throw away
-				// and start over in newL0Sublevels.
-				return errInvalidL0SublevelsOpt
+			if interval.files[len(interval.files)-1].LargestSeqNum > f.LargestSeqNum {
+				panic(errors.AssertionFailedf("addFileToSublevels found existing newer file"))
 			}
 			// interval.files is sorted by sublevels, from lowest to highest.
 			// addL0Files can only add files at sublevels higher than existing files
@@ -703,14 +688,13 @@ func (s *l0Sublevels) addFileToSublevels(f *TableMetadata, checkInvariant bool) 
 	}
 	f.SubLevel = subLevel
 	if subLevel > len(s.levelFiles) {
-		return errors.Errorf("chose a sublevel beyond allowed range of sublevels: %d vs 0-%d", subLevel, len(s.levelFiles))
+		panic(errors.AssertionFailedf("chose a sublevel beyond allowed range of sublevels: %d vs 0-%d", subLevel, len(s.levelFiles)))
 	}
 	if subLevel == len(s.levelFiles) {
 		s.levelFiles = append(s.levelFiles, []*TableMetadata{f})
 	} else {
 		s.levelFiles[subLevel] = append(s.levelFiles[subLevel], f)
 	}
-	return nil
 }
 
 func (s *l0Sublevels) calculateFlushSplitKeys(flushSplitMaxBytes int64) {
@@ -2147,8 +2131,8 @@ func (o *L0Organizer) Update(
 	}
 	// If we only added tables, try to use addL0Files.
 	if len(deletedL0Tables) == 0 {
-		newSublevels, err := o.l0Sublevels.addL0Files(addedL0Tables, o.flushSplitBytes, newLevelMeta)
-		if err == nil {
+		if files, ok := o.l0Sublevels.canUseAddL0Files(addedL0Tables, newLevelMeta); ok {
+			newSublevels := o.l0Sublevels.addL0Files(files, o.flushSplitBytes, newLevelMeta)
 			// In invariants mode, sometimes rebuild from scratch to verify that
 			// AddL0Files did the right thing. Note that NewL0Sublevels updates
 			// fields in TableMetadata like L0Index, so we don't want to do this
@@ -2169,9 +2153,6 @@ func (o *L0Organizer) Update(
 			}
 			o.l0Sublevels = newSublevels
 			return
-		}
-		if !errors.Is(err, errInvalidL0SublevelsOpt) {
-			panic(errors.AssertionFailedf("error generating L0Sublevels: %s", err))
 		}
 	}
 	var err error

--- a/internal/manifest/l0_sublevels.go
+++ b/internal/manifest/l0_sublevels.go
@@ -2073,6 +2073,7 @@ type L0Organizer struct {
 	cmp             base.Compare
 	formatKey       base.FormatKey
 	flushSplitBytes int64
+	generation      int64
 
 	// levelMetadata is the current L0.
 	levelMetadata LevelMetadata
@@ -2108,68 +2109,108 @@ func NewL0Organizer(comparer *base.Comparer, flushSplitBytes int64) *L0Organizer
 	return o
 }
 
-// SublevelFiles returns the sublevels as LevelSlices. The returned value (both
-// the slice and each LevelSlice) is immutable. The L0Organizer creates new
-// slices every time L0 changes.
-func (o *L0Organizer) SublevelFiles() []LevelSlice {
-	return o.l0Sublevels.Levels
-}
-
-// Update the L0 organizer with the given L0 changes.
-func (o *L0Organizer) Update(
-	addedL0Tables map[base.FileNum]*TableMetadata,
-	deletedL0Tables map[base.FileNum]*TableMetadata,
-	newLevelMeta *LevelMetadata,
-) {
+// PrepareUpdate is the first step in the two-step process to update the
+// L0Organizer. This first step performs as much work as it can without
+// modifying the L0Organizer.
+//
+// This method can be called concurrently with other methods (other than
+// PerformUpdate). It allows doing most of the update work outside an important
+// lock.
+func (o *L0Organizer) PrepareUpdate(bve *BulkVersionEdit, newVersion *Version) L0PreparedUpdate {
+	addedL0Tables := bve.AddedTables[0]
+	deletedL0Tables := bve.DeletedTables[0]
+	newLevelMeta := &newVersion.Levels[0]
 	if invariants.Enabled && invariants.Sometimes(10) {
 		// Verify that newLevelMeta = m.levelMetadata + addedL0Tables - deletedL0Tables.
 		verifyLevelMetadataTransition(&o.levelMetadata, newLevelMeta, addedL0Tables, deletedL0Tables)
 	}
-	o.levelMetadata = *newLevelMeta
+
 	if len(addedL0Tables) == 0 && len(deletedL0Tables) == 0 {
-		return
-	}
-	// If we only added tables, try to use addL0Files.
-	if len(deletedL0Tables) == 0 {
-		if files, ok := o.l0Sublevels.canUseAddL0Files(addedL0Tables, newLevelMeta); ok {
-			newSublevels := o.l0Sublevels.addL0Files(files, o.flushSplitBytes, newLevelMeta)
-			// In invariants mode, sometimes rebuild from scratch to verify that
-			// AddL0Files did the right thing. Note that NewL0Sublevels updates
-			// fields in TableMetadata like L0Index, so we don't want to do this
-			// every time.
-			if invariants.Enabled && invariants.Sometimes(10) {
-				expectedSublevels, err := newL0Sublevels(newLevelMeta, o.cmp, o.formatKey, o.flushSplitBytes)
-				if err != nil {
-					panic(fmt.Sprintf("error when regenerating sublevels: %s", err))
-				}
-				s1 := describeSublevels(o.formatKey, false /* verbose */, expectedSublevels.Levels)
-				s2 := describeSublevels(o.formatKey, false /* verbose */, newSublevels.Levels)
-				if s1 != s2 {
-					// Add verbosity.
-					s1 := describeSublevels(o.formatKey, true /* verbose */, expectedSublevels.Levels)
-					s2 := describeSublevels(o.formatKey, true /* verbose */, newSublevels.Levels)
-					panic(fmt.Sprintf("incremental L0 sublevel generation produced different output than regeneration: %s != %s", s1, s2))
-				}
-			}
-			o.l0Sublevels = newSublevels
-			return
+		return L0PreparedUpdate{
+			generation:   o.generation,
+			newSublevels: o.l0Sublevels,
 		}
 	}
-	var err error
-	o.l0Sublevels, err = newL0Sublevels(newLevelMeta, o.cmp, o.formatKey, o.flushSplitBytes)
+
+	if len(deletedL0Tables) == 0 {
+		if files, ok := o.l0Sublevels.canUseAddL0Files(addedL0Tables, newLevelMeta); ok {
+			return L0PreparedUpdate{
+				generation: o.generation,
+				addL0Files: files,
+			}
+		}
+	}
+	newSublevels, err := newL0Sublevels(newLevelMeta, o.cmp, o.formatKey, o.flushSplitBytes)
 	if err != nil {
 		panic(errors.AssertionFailedf("error generating L0Sublevels: %s", err))
+	}
+
+	return L0PreparedUpdate{
+		generation:   o.generation,
+		newSublevels: newSublevels,
 	}
 }
 
-// ResetForTesting reinitializes the L0Organizer to reflect a given L0 level.
-func (o *L0Organizer) ResetForTesting(levelMetadata *LevelMetadata) {
-	o.levelMetadata = *levelMetadata
+// L0PreparedUpdate is returned by L0Organizer.PrepareUpdate(), to be passed to
+// PerformUpdate().
+type L0PreparedUpdate struct {
+	generation int64
+
+	// Exactly one of the following fields will be set.
+	addL0Files   []*TableMetadata
+	newSublevels *l0Sublevels
+}
+
+// PerformUpdate applies an update the L0 organizer which was previously
+// prepared using PrepareUpdate.
+//
+// Sets newVersion.L0SublevelFiles (which is immutable once set).
+//
+// This method cannot be called concurrently with any other methods.
+func (o *L0Organizer) PerformUpdate(prepared L0PreparedUpdate, newVersion *Version) {
+	if prepared.generation != o.generation {
+		panic("invalid L0 update generation")
+	}
+	o.levelMetadata = newVersion.Levels[0]
+	o.generation++
+	if prepared.addL0Files != nil {
+		newSublevels := o.l0Sublevels.addL0Files(prepared.addL0Files, o.flushSplitBytes, &o.levelMetadata)
+		// In invariants mode, sometimes rebuild from scratch to verify that
+		// AddL0Files did the right thing. Note that NewL0Sublevels updates
+		// fields in TableMetadata like L0Index, so we don't want to do this
+		// every time.
+		if invariants.Enabled && invariants.Sometimes(10) {
+			expectedSublevels, err := newL0Sublevels(&o.levelMetadata, o.cmp, o.formatKey, o.flushSplitBytes)
+			if err != nil {
+				panic(fmt.Sprintf("error when regenerating sublevels: %s", err))
+			}
+			s1 := describeSublevels(o.formatKey, false /* verbose */, expectedSublevels.Levels)
+			s2 := describeSublevels(o.formatKey, false /* verbose */, newSublevels.Levels)
+			if s1 != s2 {
+				// Add verbosity.
+				s1 := describeSublevels(o.formatKey, true /* verbose */, expectedSublevels.Levels)
+				s2 := describeSublevels(o.formatKey, true /* verbose */, newSublevels.Levels)
+				panic(fmt.Sprintf("incremental L0 sublevel generation produced different output than regeneration: %s != %s", s1, s2))
+			}
+		}
+		o.l0Sublevels = newSublevels
+	} else {
+		o.l0Sublevels = prepared.newSublevels
+	}
+	newVersion.L0SublevelFiles = o.l0Sublevels.Levels
+}
+
+// ResetForTesting reinitializes the L0Organizer to reflect the given version.
+// Sets v.L0SublevelFiles.
+func (o *L0Organizer) ResetForTesting(v *Version) {
+	o.levelMetadata = v.Levels[0]
+	o.generation = 0
 	var err error
-	o.l0Sublevels, err = newL0Sublevels(levelMetadata, o.cmp, o.formatKey, o.flushSplitBytes)
+	o.l0Sublevels, err = newL0Sublevels(&v.Levels[0], o.cmp, o.formatKey, o.flushSplitBytes)
 	if err != nil {
 		panic(errors.AssertionFailedf("error generating L0Sublevels: %s", err))
 	}
+	v.L0SublevelFiles = o.l0Sublevels.Levels
 }
 
 // verifyLevelMetadataTransition verifies that newLevel matches oldLevel after

--- a/internal/manifest/l0_sublevels_test.go
+++ b/internal/manifest/l0_sublevels_test.go
@@ -54,9 +54,10 @@ func readManifest(filename string) (*Version, error) {
 		if err := bve.Accumulate(&ve); err != nil {
 			return nil, err
 		}
-		if v, err = bve.Apply(v, l0Organizer, 32000); err != nil {
+		if v, err = bve.Apply(v, 32000); err != nil {
 			return nil, err
 		}
+		l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, v), v)
 	}
 	return v, nil
 }

--- a/internal/manifest/l0_sublevels_test.go
+++ b/internal/manifest/l0_sublevels_test.go
@@ -19,6 +19,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/datadriven"
+	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble/internal/base"
 	"github.com/cockroachdb/pebble/internal/testkeys"
 	"github.com/cockroachdb/pebble/record"
@@ -294,10 +295,13 @@ func TestL0Sublevels(t *testing.T) {
 			levelMetadata := MakeLevelMetadata(base.DefaultComparer.Compare, 0, fileMetas[0])
 			if initialize {
 				if addL0FilesOpt {
-					sublevels, err = sublevels.addL0Files(addedL0Files, int64(flushSplitMaxBytes), &levelMetadata)
-					// Check if the output matches a full initialization.
-					sublevels2, _ := newL0Sublevels(&levelMetadata, base.DefaultComparer.Compare, base.DefaultFormatter, int64(flushSplitMaxBytes))
-					if sublevels != nil && sublevels2 != nil {
+					files, ok := sublevels.canUseAddL0Files(addedL0Files, &levelMetadata)
+					if !ok {
+						err = errors.Newf("pebble: L0 sublevel generation optimization cannot be used")
+					} else {
+						sublevels = sublevels.addL0Files(files, int64(flushSplitMaxBytes), &levelMetadata)
+						// Check if the output matches a full initialization.
+						sublevels2, _ := newL0Sublevels(&levelMetadata, base.DefaultComparer.Compare, base.DefaultFormatter, int64(flushSplitMaxBytes))
 						require.Equal(t, sublevels.flushSplitUserKeys, sublevels2.flushSplitUserKeys)
 						require.Equal(t, sublevels.levelFiles, sublevels2.levelFiles)
 					}
@@ -555,8 +559,9 @@ func TestAddL0FilesEquivalence(t *testing.T) {
 			// that of the previous L0Sublevels. So it must be called before
 			// newL0Sublevels; calling it the other way around results in
 			// out-of-bounds panics.
-			s2, err = s2.addL0Files(filesToAdd, flushSplitMaxBytes, &levelMetadata)
-			require.NoError(t, err)
+			files, ok := s2.canUseAddL0Files(filesToAdd, &levelMetadata)
+			require.True(t, ok)
+			s2 = s2.addL0Files(files, flushSplitMaxBytes, &levelMetadata)
 		}
 
 		s, err = newL0Sublevels(&levelMetadata, testkeys.Comparer.Compare, testkeys.Comparer.FormatKey, flushSplitMaxBytes)

--- a/internal/manifest/manifest_test.go
+++ b/internal/manifest/manifest_test.go
@@ -128,7 +128,8 @@ func replayManifest(
 	}
 	l0Organizer := manifest.NewL0Organizer(cmp, opts.FlushSplitBytes)
 	emptyVersion := manifest.NewInitialVersion(cmp)
-	v, err := bve.Apply(emptyVersion, l0Organizer, opts.Experimental.ReadCompactionRate)
+	v, err := bve.Apply(emptyVersion, opts.Experimental.ReadCompactionRate)
 	require.NoError(t, err)
+	l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, v), v)
 	return v, l0Organizer
 }

--- a/internal/manifest/version.go
+++ b/internal/manifest/version.go
@@ -1214,8 +1214,7 @@ func NewVersionForTesting(
 			v.Levels[l].totalSize += f.Size
 		}
 	}
-	l0Organizer.ResetForTesting(&v.Levels[0])
-	v.L0SublevelFiles = l0Organizer.SublevelFiles()
+	l0Organizer.ResetForTesting(v)
 	return v
 }
 

--- a/internal/manifest/version_edit.go
+++ b/internal/manifest/version_edit.go
@@ -1149,9 +1149,10 @@ func (b *BulkVersionEdit) Accumulate(ve *VersionEdit) error {
 // the levels.
 //
 // curr may be nil, which is equivalent to a pointer to a zero version.
-func (b *BulkVersionEdit) Apply(
-	curr *Version, l0Organizer *L0Organizer, readCompactionRate int64,
-) (*Version, error) {
+//
+// Not that L0SublevelFiles is not initialized in the returned version; it is
+// the caller's responsibility to set it using L0Organizer.PerformUpdate().
+func (b *BulkVersionEdit) Apply(curr *Version, readCompactionRate int64) (*Version, error) {
 	comparer := curr.cmp
 	v := &Version{
 		cmp: comparer,
@@ -1292,9 +1293,6 @@ func (b *BulkVersionEdit) Apply(
 			}
 		}
 	}
-
-	l0Organizer.Update(b.AddedTables[0], b.DeletedTables[0], &v.Levels[0])
-	v.L0SublevelFiles = l0Organizer.SublevelFiles()
 
 	// We maintain stats about active references in blob files and can infer
 	// when a blob file has become a 'zombie,' and is no longer referenced in

--- a/internal/manifest/version_edit_test.go
+++ b/internal/manifest/version_edit_test.go
@@ -449,10 +449,11 @@ func TestVersionEditApply(t *testing.T) {
 				if l0Organizer == nil {
 					d.Fatalf(t, "no L0 organizer")
 				}
-				newv, err := bve.Apply(v, l0Organizer, readCompactionRate)
+				newv, err := bve.Apply(v, readCompactionRate)
 				if err != nil {
 					return err.Error()
 				}
+				l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, newv), newv)
 				if saveName != "" {
 					versions[saveName] = newv
 					l0Organizers[saveName] = l0Organizer
@@ -461,7 +462,7 @@ func TestVersionEditApply(t *testing.T) {
 				// Reinitialize the L0 organizer in case we want to use the same version
 				// again (l0Organizer now reflects newv).
 				l0Organizer = NewL0Organizer(base.DefaultComparer, flushSplitBytes)
-				l0Organizer.ResetForTesting(&v.Levels[0])
+				l0Organizer.ResetForTesting(v)
 				l0Organizers[name] = l0Organizer
 
 				return newv.DebugString()

--- a/replay/replay.go
+++ b/replay/replay.go
@@ -802,10 +802,11 @@ func (r *Runner) prepareWorkloadSteps(ctx context.Context) error {
 				}
 
 				// Apply the edit.
-				v, err = bve.Apply(v, l0Organizer, r.Opts.Experimental.ReadCompactionRate)
+				v, err = bve.Apply(v, r.Opts.Experimental.ReadCompactionRate)
 				if err != nil {
 					return err
 				}
+				l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, v), v)
 				// AddedTablesByFileNum maps file number to table metadata for all added
 				// sstables from accumulated version edits so we must retain it.
 				bve = manifest.BulkVersionEdit{AddedTablesByFileNum: bve.AddedTablesByFileNum}

--- a/tool/db.go
+++ b/tool/db.go
@@ -723,10 +723,11 @@ func (d *dbT) runProperties(cmd *cobra.Command, args []string) {
 		}
 		l0Organizer := manifest.NewL0Organizer(cmp, d.opts.FlushSplitBytes)
 		emptyVersion := manifest.NewInitialVersion(cmp)
-		v, err := bve.Apply(emptyVersion, l0Organizer, d.opts.Experimental.ReadCompactionRate)
+		v, err := bve.Apply(emptyVersion, d.opts.Experimental.ReadCompactionRate)
 		if err != nil {
 			return err
 		}
+		l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, v), v)
 
 		objProvider, err := objstorageprovider.Open(objstorageprovider.DefaultSettings(d.opts.FS, dirname))
 		if err != nil {

--- a/tool/manifest.go
+++ b/tool/manifest.go
@@ -243,11 +243,12 @@ func (m *manifestT) runDump(cmd *cobra.Command, args []string) {
 			if comparer != nil {
 				l0Organizer := manifest.NewL0Organizer(comparer, 0 /* flushSplitBytes */)
 				emptyVersion := manifest.NewInitialVersion(comparer)
-				v, err := bve.Apply(emptyVersion, l0Organizer, m.opts.Experimental.ReadCompactionRate)
+				v, err := bve.Apply(emptyVersion, m.opts.Experimental.ReadCompactionRate)
 				if err != nil {
 					fmt.Fprintf(stdout, "%s\n", err)
 					return
 				}
+				l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, v), v)
 				m.printLevels(comparer.Compare, stdout, v)
 			}
 		}()
@@ -628,7 +629,7 @@ func (m *manifestT) runCheck(cmd *cobra.Command, args []string) {
 					l0Organizer = manifest.NewL0Organizer(cmp, 0 /* flushSplitBytes */)
 					v = manifest.NewInitialVersion(cmp)
 				}
-				newv, err := bve.Apply(v, l0Organizer, m.opts.Experimental.ReadCompactionRate)
+				newv, err := bve.Apply(v, m.opts.Experimental.ReadCompactionRate)
 				if err != nil {
 					fmt.Fprintf(stdout, "%s: offset: %d err: %s\n",
 						arg, offset, err)
@@ -648,6 +649,7 @@ func (m *manifestT) runCheck(cmd *cobra.Command, args []string) {
 					ok = false
 					break
 				}
+				l0Organizer.PerformUpdate(l0Organizer.PrepareUpdate(&bve, newv), newv)
 				v = newv
 			}
 		}()

--- a/version_set.go
+++ b/version_set.go
@@ -342,10 +342,11 @@ func (vs *versionSet) load(
 	}
 
 	emptyVersion := manifest.NewInitialVersion(opts.Comparer)
-	newVersion, err := bve.Apply(emptyVersion, vs.l0Organizer, opts.Experimental.ReadCompactionRate)
+	newVersion, err := bve.Apply(emptyVersion, opts.Experimental.ReadCompactionRate)
 	if err != nil {
 		return err
 	}
+	vs.l0Organizer.PerformUpdate(vs.l0Organizer.PrepareUpdate(&bve, newVersion), newVersion)
 	vs.l0Organizer.InitCompactingFileInfo(nil /* in-progress compactions */)
 	vs.append(newVersion)
 
@@ -561,6 +562,7 @@ func (vs *versionSet) logAndApply(
 	zombieBackings, removedVirtualBackings, localLiveSizeDelta :=
 		getZombiesAndUpdateVirtualBackings(ve, &vs.virtualBackings, vs.provider)
 
+	var l0Update manifest.L0PreparedUpdate
 	if err := func() error {
 		vs.mu.Unlock()
 		defer vs.mu.Lock()
@@ -568,17 +570,16 @@ func (vs *versionSet) logAndApply(
 		if vs.getFormatMajorVersion() < FormatVirtualSSTables && len(ve.CreatedBackingTables) > 0 {
 			return base.AssertionFailedf("MANIFEST cannot contain virtual sstable records due to format major version")
 		}
-		var b bulkVersionEdit
-		err := b.Accumulate(ve)
+		var bulkEdit bulkVersionEdit
+		err := bulkEdit.Accumulate(ve)
 		if err != nil {
 			return errors.Wrap(err, "MANIFEST accumulate failed")
 		}
-		newVersion, err = b.Apply(
-			currentVersion, vs.l0Organizer, vs.opts.Experimental.ReadCompactionRate,
-		)
+		newVersion, err = bulkEdit.Apply(currentVersion, vs.opts.Experimental.ReadCompactionRate)
 		if err != nil {
 			return errors.Wrap(err, "MANIFEST apply failed")
 		}
+		l0Update = vs.l0Organizer.PrepareUpdate(&bulkEdit, newVersion)
 
 		if newManifestFileNum != 0 {
 			if err := vs.createManifest(vs.dirname, newManifestFileNum, minUnflushedLogNum, nextFileNum, newManifestVirtualBackings); err != nil {
@@ -642,6 +643,7 @@ func (vs *versionSet) logAndApply(
 	// L0Sublevels.
 	inProgress := inProgressCompactions()
 
+	vs.l0Organizer.PerformUpdate(l0Update, newVersion)
 	vs.l0Organizer.InitCompactingFileInfo(inProgressL0Compactions(inProgress))
 
 	// Update the zombie tables set first, as installation of the new version

